proto/skaffold.proto

```diff
@@ -104,9 +112,33 @@ message Event {
         ResourceStatusCheckEvent resourceStatusCheckEvent = 6; // indicates progress for each kubernetes deployment.
         FileSyncEvent fileSyncEvent = 7; // describes the sync status.
         DebuggingContainerEvent debuggingContainerEvent = 8; // describes the appearance or disappearance of a debugging container
+        DevLoopEvent devLoopEvent = 9; // describes a start and end of a dev loop.
+        EndEvent endEvent = 10;
     }
 }
 
+// `EndEvent` is the last event indicating skaffold session ended.
+message EndEvent {
+    ErrorCode errCode = 1;
+    string errMsg = 2;
+    repeated DevLoop loops = 3;
+}
+// `DevLoopEvent` marks the start and end of a dev loop.
+message DevLoopEvent {
+    int32 iteration = 1; // dev loop iteration
+    ChangeType changeType = 2;
+    string status = 3; // dev loop status oneof: In Progress, Completed, Failed
+    ActionableErr err = 4; // actionable error message
+
+}
+
+// `ErrDef` defines an error occurred along with an optional suggestions
+message ActionableErr {
+    ErrorCode errCode = 1; // error code representing the error
+    string message = 2; // message describing the error.
+    repeated string suggestions = 3; // a list of suggestions for the error.
+
```

Comment:
I don't think reporting messages here is the right thing to do.  I'm wondering if it would be better to have a secondary set of possible action codes that could be returned instead.  That would also allow i18n or more.

(I think it's also an opportunity to somehow segregate PII fields, like artifact names or image tags, so that we might be able to report the actual errors seen?)

Reply:
Thanks. I should have made this clear. The suggestion text is just for IDEs to show to User.
Having suggestions plumbed through skaffold will reduce the duplication of efforts.
I will add that as a comment.
For metrics, we should probably report codes like you mentioned.